File: app/src/main/java/com/inqint/yarnrequirements/ProjectRecyclerViewAdapter.java
package com.inqint.yarnrequirements;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.inqint.yarnrequirements.ProjectListFragment.OnListFragmentInteractionListener;
import com.inqint.yarnrequirements.Projects.Project;

import java.util.List;

/**
 * {@link RecyclerView.Adapter} that can display a {@link Project} and makes a call to the
 * specified {@link OnListFragmentInteractionListener}.
 */
public class ProjectRecyclerViewAdapter extends RecyclerView.Adapter<ProjectRecyclerViewAdapter.ViewHolder> {

    private final List<Project> mValues;
    private final OnListFragmentInteractionListener mListener;

    public ProjectRecyclerViewAdapter(List<Project> items, OnListFragmentInteractionListener listener) {
        mValues = items;
        mListener = listener;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.fragment_project_icon, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, int position) {
        holder.mItem = mValues.get(position);
        holder.mNameView.setText(holder.mItem.getName());
        holder.imageView.setImageResource(holder.mItem.getThumbImageID());

        holder.mView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (null != mListener) {
                    // Notify the active callbacks interface (the activity, if the
                    // fragment is attached to one) that an item has been selected.
                    mListener.onListFragmentInteraction(holder.mItem);
                }
            }
        });
    }

    @Override
    public int getItemCount() {
        return mValues.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        public final View mView;
        public final TextView mNameView;
        public final ImageView imageView;
        public Project mItem;

        public ViewHolder(View view) {
            super(view);
            mView = view;
            mNameView = (TextView) view.findViewById(R.id.name);
            imageView = (ImageView) view.findViewById(R.id.thumb_image);
        }

        @Override
        public String toString() {
            return super.toString() + " '" + mNameView.getText() + "'";
        }
    }

}

File: app/src/main/java/com/inqint/yarnrequirements/ProjectRecyclerViewAdapter.kt
package com.inqint.yarnrequirements

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.navigation.findNavController
import androidx.recyclerview.widget.RecyclerView
import com.inqint.yarnrequirements.Projects.DimensionProject
import com.inqint.yarnrequirements.Projects.Project
import com.inqint.yarnrequirements.Projects.SizeProject

class ProjectRecyclerViewAdapter(
    private val values: List<ProjectContent.FragmentProject>
) : RecyclerView.Adapter<ProjectRecyclerViewAdapter.ViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.fragment_project_icon, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.item = values[position].project
        holder.nameView.text = holder.item!!.name
        holder.imageView.setImageResource(holder.item!!.thumbImageID)

        holder.mView.setOnClickListener {
            val project = values[position].project
            if (project is DimensionProject) {
                val projectFragment = ProjectListFragmentDirections.actionListToDimensionProjectFragment(project)
                it.findNavController().navigate(projectFragment)
            }
            else if (project is SizeProject) {
                val projectFragment = ProjectListFragmentDirections.actionListToSizeProject(project)
                it.findNavController().navigate(projectFragment)
            }
            else {
                val projectFragment = ProjectListFragmentDirections.actionListToSockProjectFragment(project)
                it.findNavController().navigate(projectFragment)
            }
        }
    }

    override fun getItemCount(): Int {
        return values.size
    }

    inner class ViewHolder(val mView: View) : RecyclerView.ViewHolder(mView) {
        val nameView = mView.findViewById<View>(R.id.name) as TextView
        val imageView: ImageView
        var item: Project? = null

        init {
            imageView = mView.findViewById<View>(R.id.thumb_image) as ImageView
        }

        override fun toString(): String {
            return super.toString() + " '" + nameView.text + "'"
        }
    }

}
